x/upgrade/abci.go

```diff
@@ -3,22 +3,34 @@ package upgrade
 import (
 	"fmt"
 
+	"github.com/spf13/viper"
 	abci "github.com/tendermint/tendermint/abci/types"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
 // BeginBlock will check if there is a scheduled plan and if it is ready to be executed.
+// If skip upgrade flag is set, it will skip and clear the upgrade plan
 // If it is ready, it will execute it if the handler is installed, and panic/abort otherwise.
 // If the plan is not ready, it will ensure the handler is not registered too early (and abort otherwise).
 //
-// The prupose is to ensure the binary is switch EXACTLY at the desired block, and to allow
+// The purpose is to ensure the binary is switched EXACTLY at the desired block, and to allow
 // a migration to be executed if needed upon this switch (migration defined in the new binary)
 func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
+
 	plan, found := k.GetUpgradePlan(ctx)
 	if !found {
 		return
 	}
+
+	skipUpgrade := viper.Get(FlagUnsafeSkipUpgrade).(bool)
+	if skipUpgrade {
```

Comment:
This logic looks proper. However, I think we need one more condition. All nodes must execute this `ClearUpgradePlan` at the same block.

The solution I can think of is to place this `if skipUpgrade` block inside `if plan.ShouldExecute()` - at the top. Thus, when it comes time to execute, it will clear the plan instead and continue, using the came coordination logic we use for applying the upgrade.

Reply:
Yes,changed accordingly.